### DISCUSSION TITLE
Improve `eglot-signature-eldoc-function`: More Helpful Documentation!

### DISCUSSION THREAD

**Discussion by mekeor** [joaotavora/eglot#1200] [Ideas]
Hello. I wrote a custom function for `eldoc-documentation-functions`, meant as a replacement for `eglot-signature-eldoc-function`. As far as I understand, the [chances that Eglot shows any signature-documentation at all](https://github.com/joaotavora/eglot/blob/dc2efa638892d521cb456a591051e2a58c785bb1/eglot.el#L3074-L3076), are very very small, unfortunately. I think we should improve that.

The screenshots show: The original function often just lists loads of type signatures. The customized function shows the type signature and documentation for the currently active "signature".

![eglot-original](https://user-images.githubusercontent.com/60295641/229445634-0de9b93c-8ced-4f04-a625-331f3228b549.png)

![eglot-custom](https://user-images.githubusercontent.com/60295641/229444163-bb04ddbb-9d27-4e62-840e-32b8ecc82e1d.png)

Here is the code I use:

```emacs-lisp
(defun /eglot-signature-information (cb)
  (when (eglot--server-capable :signatureHelpProvider)
    (let ((buf (current-buffer)))
      (jsonrpc-async-request
        (eglot--current-server-or-lose)
        :textDocument/signatureHelp
        (eglot--TextDocumentPositionParams)
        :success-fn
        (eglot--lambda
          ((SignatureHelp)
            signatures activeSignature activeParameter)
          (let
            ((targetSignatures
               (if activeSignature
                 (list (seq-elt signatures activeSignature))
                 signatures)))
            (eglot--when-buffer-window buf
              (funcall cb
                (unless (seq-empty-p targetSignatures)
                  (string-join
                    (seq-mapcat
                      (eglot--lambda
                        ((SignatureInformation)
                          label documentation
                          parameters activeParameter)
                        (unless (null documentation)
                          (list label
                            (eglot--format-markup documentation))))
                      targetSignatures)
                    "\n\n"))))))
        :deferred :textDocument/signatureHelp))
    t))

(defun /eldoc-add-signature-information ()
  (setq eldoc-documentation-strategy
    #'eldoc-documentation-compose)
  (setq-local eldoc-documentation-functions
    (list
      #'/eglot-signature-information
      #'eglot-hover-eldoc-function)))

(add-hook 'eglot-managed-mode-hook #'/eldoc-add-signature-information)
```

PS: Sorry for not capturing the cursor in the second screenshot. Just imagine it. It's there, but it's blinking.

**Comment by joaotavora** [comment]
Hi @mekeor, i did changes to this very recently. If you can, look at the version of Eglot in Emacs master as of today and tell me what you think. I think we had the same idea, more or less.

**Reply by mekeor** [reply]
Wow, thanks for the blazingly fast reply. I'm using Emacs' master version anyway. So, let me read the new code and test it. I'm excited! :)

**Reply by mekeor** [reply]
The news sounds great, too: https://github.com/emacs-mirror/emacs/blob/master/etc/EGLOT-NEWS#L42-L47

> Eglot takes advantage of new features of ElDoc to separate short documentation strings from large ones, sending the former to be shown in the ElDoc's echo area and the latter to be shown in other outlets, such as the *eldoc* buffer obtainable with 'C-h .'.

**Reply by mekeor** [reply]
One difference seems to be that you use the documentation from ParameterInformation, whereas I used the SignatureInformation. I'm excited to see if your solution works for me.

**Reply by joaotavora** [reply]
I don't follow what you mean, but that's OK.  I didn't give this as much testing as I would have liked across multiple servers, etc.  So I'm looking forward to feedback and we can make small adjustments based on that for the next version.

**Reply by joaotavora** [reply]
You can now try these features in Eglot 1.14

**Reply by mekeor** [reply]
I was assuming it depends on other code changes in Emacs' master branch because it uses new features of Eldoc.

**Reply by joaotavora** [reply]
And it does, but those changes are also published as packages

**Reply by mekeor** [reply]
I currently can't use Emacs from master because of this bug: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=62481

**Reply by joaotavora** [reply]
Well, you don't need to.  You can try Eglot 1.14 from any version 28, 29 (it's broken on 26 and 27 unfortunately, see #1201, but easy enough to workaround)

**Reply by mekeor** [reply]
(I would be very happy if you could merge emacs-29 into master, so the bug resolves. But sorry to mention all of this here, as it's rather off-topic.)

**Reply by mekeor** [reply]
I tried out Eglot 1.14 on Emacs-30, commit ac7ec87a7a0db887e4ae7fe9005aea517958b778. But I would love to try out the new features of Eldoc, too.

**Reply by joaotavora** [reply]
> (I would be very happy if you could merge emacs-29 into master, so the bug resolves. But sorry to mention all of this here, as it's rather off-topic.)

I did this two days ago, but I don't know what you are talking about, so ping me in the context of whatever bug this is related to, if you think I can be of help.

> I tried out Eglot 1.14 on Emacs-30, commit ac7ec87a7a0db887e4ae7fe9005aea517958b778. But I would love to try out the new features of Eldoc, too.

In that case, you have experienced the "new features of ElDoc" already :-) , and if nothing broke, then that's what's supposed to happen.  Just one of these features is user-visible, and quite discreet: the separating line in the `*eldoc*` buffer has been changed and can be customized with `eldoc-buffer-separator` or something like that.

**Reply by mekeor** [reply]
No, I did not experience the "new features of ElDoc" already. E.g. in my version of Emacs, there still is no `:echo` field for `eldoc-documentation-functions`. The commit I mentioned, [ac7ec87a7a0db887e4ae7fe9005aea517958b778](https://github.com/emacs-mirror/emacs/commit/ac7ec87a7a0db887e4ae7fe9005aea517958b778), is from 3rd February 2023. You added some Eldoc features afterwards, AFAIK.

So, without having the `:echo` feature, of course, my mini-buffer is still quite flooded with Eglot 1.14. (I.e. all Emacs users who don't use Emacs version 30 will continue to suffer from a flooded mini-buffer.)

![2023-04-05-003842_1280x800_scrot](https://user-images.githubusercontent.com/60295641/229938062-07d0f5a6-ddef-47e3-b504-ff8a05b68b44.png)

**Reply by joaotavora** [reply]
> So, without having the :echo feature, of course, my mini-buffer is still quite flooded with Eglot 1.14. (I.e. all Emacs users who don't use Emacs version 30 will continue to suffer from a flooded mini-buffer.)

I don't understand, Eglot 1.14 depends on Eldoc 1.14.0, and that does have the `:echo` cookie feature. Can you check if you have it?

**Reply by mekeor** [reply]
Here's the relevant code:

```
             (if (not active-sig) (funcall cb nil)
               (funcall cb
                        (mapconcat #'eglot--sig-info signatures "\n")
```

That means, Eglot shows the documentation of each and every signature, even if there is an active signature. Why not show the documentation only for the active signature?

Also: When there is an active parameter and the parameter's documentation (`((ParameterInformation) label documentation)`) is nil, we should still show the signature's documentation (`((SignatureInformation) label documentation parameters activeParameter)`).

**Reply by mekeor** [reply]
> > So, without having the :echo feature, of course, my mini-buffer is still quite flooded with Eglot 1.14. (I.e. all Emacs users who don't use Emacs version 30 will continue to suffer from a flooded mini-buffer.)
> 
> I don't understand, Eglot 1.14 depends on Eldoc 1.14.0, and that does have the `:echo` cookie feature. Can you check if you have it?

Oh. To be honest, I just evaluated the Eglot 1.14 source code. Sorry. I will do it correctly.

**Reply by mekeor** [reply]
Latest Emacs master branch still does not build with native-compilation. (https://debbugs.gnu.org/cgi/bugreport.cgi?bug=62537)

I will now build it without native-compilation.

Until then, feel free to test my code, if you want:

```el
(defun eglot-signature-eldoc-function (cb)
  (when (eglot--server-capable :signatureHelpProvider)
    (let ((buf (current-buffer)))
      (jsonrpc-async-request
        (eglot--current-server-or-lose)
        :textDocument/signatureHelp
        (eglot--TextDocumentPositionParams)
        :success-fn
        (eglot--lambda
          ((SignatureHelp)
            signatures activeSignature activeParameter)
          (eglot--when-buffer-window buf
            (let
              ((signatures
                 (if activeSignature
                   (list (seq-elt signatures activeSignature))
                   signatures))
                (sig-help-active-param activeParameter))
              (funcall cb
                (unless (seq-empty-p signatures)
                  (string-join
                    (seq-mapcat
                      (eglot--lambda
                        ((SignatureInformation)
                          label documentation
                          parameters activeParameter)
                        (let
                          ((active-param (or activeParameter
                                           sig-help-active-param)))
                          (if (or (null active-param)
                                (< active-param 0)
                                (<= (length parameters) active-param))
                            (unless (null documentation)
                              (list label
                                (eglot--format-markup
                                  documentation)))
                            (let
                              ((sig-info-label label)
                                (sig-info-documentation documentation))
                              (eglot--dbind
                                ((ParameterInformation)
                                  label documentation)
                                (seq-elt parameters active-param)
                                (let ((param-info-label-formatted
                                        (if (stringp label)
                                          label
                                          ;; TODO: "inclusive start
                                          ;; and exclusive end
                                          ;; offsets within its
                                          ;; containing signature
                                          ;; label"
                                          nil)))
                                  (if documentation
                                    (list param-info-label-formatted
                                      (eglot--format-markup
                                        documentation)))
                                  (list param-info-label-formatted
                                    sig-info-label
                                    (eglot--format-markup
                                      sig-info-documentation))))))))
                      signatures)
                    "\n\n"))))))
        :deferred :textDocument/signatureHelp))
    t))
```

**Reply by mekeor** [reply]
Here's a screenshot based on my code.

![2023-04-05-015313_1280x800_scrot](https://user-images.githubusercontent.com/60295641/229947307-fd7f3536-4172-4dcd-a53a-69b17dce5dad.png)

**Reply by mekeor** [reply]
I installed latest Emacs from master, loaded Eglot 1.14 and it out. The mini-buffer is still flooded for me. There is no need for a new screenshot. It remains all the same.

**Reply by joaotavora** [reply]
> I installed latest Emacs from master, loaded Eglot 1.14

If you installed "latest" (what commit exactly?) Emacs, then you don't need to do that -- whatever "loaded" means (could mean `package-install`, could mean `require`, or even `load`)

Regardless, I think it's time for you to show a reproducible recipe from the beginning, exactly as described in as there are too many unknown variables in your configuration/setup.  I don't see any "flooding" neither with Emacs 29 + the GNU ELPA new packages or with Emacs 30.  So please describe your problem in a new conversation or a, `M-x report-emacs-bug` according to the instructions in https://joaotavora.github.io/eglot/#Troubleshooting-Eglot

**Reply by mekeor** [reply]
I'm very sorry for all the irritation. I now fixed the issues with my local setup. I had another package installed which was pulling in Eglot 1.13. Please allow me one more attempt to rescue this conversation.

TLDR: I have now installed latest Emacs from master (probably commit 830bf66dadbb4b9b9393ade53b890b5497bdf479, but I don't know exactly, sorry), including Eglot 1.14; and I don't have any other Eglot version installed.

I can indeed reproduce your claim that there is no flooding with default Eglot settings. Eglot 1.14 on Emacs-master just shows one type signature for me. (And also some hover-information, but let's just focus on `eglot-signature-eldoc-function` in this conversation.) That's nice.

What I suggest is to not only echo that single type signature (which is the value of `sig-info-label`), but also to show `param-info-label` and `sig-info-documentation` whenever they are available.

Below are two screenshots. The first shows the default Eglot settings where only `sig-info-label` is echoed. The second screenshot shows what my version of `eglot-signature-eldoc-function` echoes: `param-info-label`, `sig-info-label` and `sig-info-documentation`. (Both screenshots also show one line from `eglot-hover-eldoc-function`, namely `var undefined`.)

![2023-04-05-122533_1280x800_scrot](https://user-images.githubusercontent.com/60295641/230057391-45920138-3cc4-4dfc-9996-cf383fab82b7.png)
![2023-04-05-122633_1280x800_scrot](https://user-images.githubusercontent.com/60295641/230057397-1fa3d30d-07ed-4b4c-98b0-742b77068c58.png)

Here's the latest version of my code:

```el
(defun eglot-signature-eldoc-function (cb)
  (when (eglot--server-capable :signatureHelpProvider)
    (let ((buf (current-buffer)))
      (jsonrpc-async-request
        (eglot--current-server-or-lose)
        :textDocument/signatureHelp
        (eglot--TextDocumentPositionParams)
        :success-fn
        (eglot--lambda
          ((SignatureHelp)
            signatures activeSignature activeParameter)
          (eglot--when-buffer-window buf
            (let
              ((signatures
                 (if activeSignature
                   (list (seq-elt signatures activeSignature))
                   signatures))
                (sig-help-active-param activeParameter))
              (funcall cb
                (unless (seq-empty-p signatures)
                  (string-join
                    (seq-mapcat
                      (eglot--lambda
                        ((SignatureInformation)
                          label documentation
                          parameters activeParameter)
                        (let
                          ((active-param (or activeParameter
                                           sig-help-active-param)))
                          (if (or (null active-param)
                                (< active-param 0)
                                (<= (length parameters) active-param))
                            (unless (null documentation)
                              (list label
                                (eglot--format-markup
                                  documentation)))
                            (let
                              ((sig-info-label label)
                                (sig-info-documentation
                                  documentation))
                              (eglot--dbind
                                ((ParameterInformation)
                                  label documentation)
                                (seq-elt parameters active-param)
                                (let ((param-info-label-formatted
                                        (if (stringp label)
                                          (list label)
                                          ;; TODO: label can also be
                                          ;; an pair of numbers
                                          nil)))
                                  (if documentation
                                    (append param-info-label-formatted
                                      (list (eglot--format-markup
                                        documentation)))
                                    (append
                                      param-info-label-formatted
                                      (list sig-info-label)
                                      (unless
                                        (null sig-info-documentation)
                                        (list
                                          (eglot--format-markup
                                            sig-info-documentation)))))))))))
                      signatures)
                    "\n\n"))))))
        :deferred :textDocument/signatureHelp))
    t))
```

My code will either show `param-info-label` and `param-info-documentation`; or, if latter is not available, it will show `param-info-label`, `sig-info-label` and `sig-info-documentation`.

**Reply by mekeor** [reply]
From what I understand, Eglot never shows the full `sig-info-documentation`. Instead, it tries to extract the last called function out of `a().chain().like().this()` and then only attempts to show `this()` or so:

```el
        ;; Decide whether to add one-line-summary to signature line
        (when (and (stringp documentation)
                   (string-match "[[:space:]]*\\([^.\r\n]+[.]?\\)"
                                 documentation))
          (setq documentation (match-string 1 documentation))
          (unless (string-prefix-p (string-trim documentation) label)
            (goto-char (point-max))
            (insert ": " (eglot--format-markup documentation))))
```

**Reply by joaotavora** [reply]
> I'm very sorry for all the irritation. I now fixed the issues with my local setup. I had another package installed which was pulling in Eglot 1.13

OK, that explains it. But there was no irritation, just perplexity at your reports.

I don't have the resources to experiment with a full replacement of an eglot function  and calculate all the implications. I don't understand what's wrong with the latest version we're discussion, which doesn't meant there isn't anything wrong, only that this description is becoming very confusing.

So I suggest that you `M-x report-emacs-bug` with a good description of what you see right now in a given specific, perfectly-reproducible situation (read the instructions I linked closely), and what you would like to see instead.  Alternatively, submit a patch.  The smaller the better. Then I can evaluate your patch against my set of tests.

**Reply by mekeor** [reply]
For the record, here's the bug report: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=62687

**Reply by joaotavora** [reply]
OK, I see it let's follow up there.

**Reply by joaotavora** [reply]
Oh, _please_ don't send back-to-back emails to `bug-gnu-emacs@gnu.org`.  That just makes the bug tracker create more bugs.  Wait for the confirmation of the bug tracker. You've sent two mails already, wait a bit until the moderators approve them.

**Reply by mekeor** [reply]
Oh nooo. Damnit!

**Reply by mekeor** [reply]
Hmm, I don't see any accidentally created bug reports when I search my name in Emacs' bug tracker:

https://debbugs.gnu.org/cgi/pkgreport.cgi?package=emacs;include=originator%3Amekeor

All my e-mails are in one single bug thread.

**Reply by mekeor** [reply]
On the bug-gnu-emacs mailing-list, my e-mails also appear in one single thread: https://lists.gnu.org/archive/html/bug-gnu-emacs/2023-04/msg00269.html

**Comment by mekeor** [comment]
On April 4, 2023 11:45:18 AM GMT+02:00, "João Távora" ***@***.***> wrote:


>I did this two days ago, but I don't know what you are talking about, so ping me in the context of whatever bug this is related to, if you think I can be of help.

Nevermind! Eli had claimed that the bug that I'm experiencing, has been resolved on emacs-29, but apparently the bug on the master-branch is different and still outstanding. So, you can't help me right now.

https://debbugs.gnu.org/cgi/bugreport.cgi?bug=62596#15

>In that case, you have experienced the "new features of ElDoc" already :-) , and if nothing broke, then that's what's supposed to happen.  Just one of these features is user-visible, and quite discreet: the separating line in the `*eldoc*` buffer has been changed and can be customized with `eldoc-buffer-separator` or something like that.

Alright then. I'll give some feedback soon.

**Comment by joaotavora** [comment]
Let's hope you haven't created 3 bug reports. Maybe the tracker is smart.
But if you have, it's no tragedy, just give the moderators some time.

João

**Reply by mekeor** [reply]
Quoting debbugs.gnu.org:

>  To reply to an existing report, for example #123, send mail to 123@debbugs.gnu.org. Or to the project's bug mailing list address (e.g. bug-gnu-emacs), but be sure to keep "bug#123" in the subject, else you will create a new report.

So, no need to worry.
